app/driver/i2c_master.c

```diff
@@ -77,86 +66,170 @@ i2c_master_setDC(uint8 SDA, uint8 SCL)
     } else {
         I2C_MASTER_SDA_HIGH_SCL_HIGH();
     }
-    if(1 == SCL) { //clock stretching
+    if(1 == SCL) {
         do {
-        } while(0 == GPIO_INPUT_GET(GPIO_ID_PIN(I2C_MASTER_SCL_GPIO)));
+            sclLevel = GPIO_INPUT_GET(GPIO_ID_PIN(I2C_MASTER_SCL_GPIO));
+        } while(sclLevel == 0);
     }
+    i2c_master_wait(5);
+#else
+    uint32 cycles_start;
+    uint32 cycles_curr;
+    uint32 this_SDA_SCL_set_mask = (SDA << i2c[id].pinSDA) | (SCL << i2c[id].pinSCL);
+    uint32 this_SDA_SCL_clear_mask = i2c[id].pin_SDA_SCL_mask ^ this_SDA_SCL_set_mask;
+    i2c[id].m_nLastSDA = SDA;
+    i2c[id].m_nLastSCL = SCL;
+
+    if(i2c[id].cycles_delay > 0){//
+        asm volatile("rsr %0, ccount":"=a"(cycles_start)); //Cycle count register
+        do{
+            asm volatile("rsr %0, ccount":"=a"(cycles_curr));
+        } while (cycles_curr - cycles_start <  i2c[id].cycles_delay);
+    }
+    if(16 == i2c[id].pinSCL){ //if GPIO16
+        WRITE_PERI_REG(RTC_GPIO_OUT,
+                       (READ_PERI_REG(RTC_GPIO_OUT) & (uint32)GPIO16_RTC_REG_MASK) | (uint32)SCL);
+        GPIO_REG_WRITE(GPIO_OUT_W1TC_ADDRESS, this_SDA_SCL_clear_mask); //SCL is shifted out of mask so no worry
+        GPIO_REG_WRITE(GPIO_OUT_W1TS_ADDRESS, this_SDA_SCL_set_mask);
+    }
+    else{
+        GPIO_REG_WRITE(GPIO_OUT_W1TC_ADDRESS, this_SDA_SCL_clear_mask);
+        GPIO_REG_WRITE(GPIO_OUT_W1TS_ADDRESS, this_SDA_SCL_set_mask);
+        if(1 == SCL) { //clock stretching
+            //instead of: while(!((gpio_input_get() >> pinSCL) & 1)) {}
```

Comment:
Again out-of-line this.  Also explain in the inline function comments that you've replicated the code of  gpio_input_get() to give a tighter timing loop.

Reply:
Done this.